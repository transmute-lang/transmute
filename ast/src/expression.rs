use crate::literal::Literal;
use crate::operators::{BinaryOperator, UnaryOperator};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
}

impl Expression {
    pub fn new(id: ExprId, kind: ExpressionKind, span: Span) -> Self {
        Self { id, kind, span }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Binary(ExprId, BinaryOperator, ExprId),
    Unary(UnaryOperator, ExprId),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    /// A dummy expression kind, inserted by the parser when the expression could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}
