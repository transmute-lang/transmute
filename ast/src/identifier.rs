use transmute_core::ids::IdentId;
use transmute_core::span::Span;

/// Represents an identifier when not used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct Identifier {
    id: IdentId,
    span: Span,
}

impl Identifier {
    pub fn new(id: IdentId, span: Span) -> Self {
        Self { id, span }
    }

    pub fn id(&self) -> IdentId {
        self.id
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn take_span(self) -> Span {
        self.span
    }
}
