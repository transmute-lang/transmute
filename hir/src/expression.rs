use crate::literal::Literal;
use crate::typed::{Typed, TypedState, Untyped};
use std::fmt::Debug;
use transmute_ast::expression::Expression as AstExpression;
use transmute_ast::expression::ExpressionKind as AstExpressionKind;
use transmute_ast::expression::Target as AstTarget;
use transmute_core::ids::{ExprId, IdentRefId, StmtId, TypeId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression<T>
where
    T: TypedState,
{
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
    typed: T,
}

impl Expression<Untyped> {
    pub fn typed(self, ty: TypeId) -> Expression<Typed> {
        Expression {
            id: self.id,
            kind: self.kind,
            span: self.span,
            typed: Typed(ty),
        }
    }
}

impl From<AstExpression> for Expression<Untyped> {
    fn from(value: AstExpression) -> Self {
        Self {
            id: value.id,
            span: value.span.clone(),
            kind: match value.kind {
                AstExpressionKind::Assignment(target, expr_id) => {
                    ExpressionKind::Assignment(Target::from(target), expr_id)
                }
                AstExpressionKind::If(cond, true_branch, false_branch) => {
                    ExpressionKind::If(cond, true_branch, false_branch)
                }
                AstExpressionKind::Literal(literal) => {
                    ExpressionKind::Literal(Literal::from(literal))
                }
                AstExpressionKind::Access(expr_id, ident_ref_id) => {
                    ExpressionKind::Access(expr_id, ident_ref_id)
                }
                AstExpressionKind::FunctionCall(ident_ref_id, expr_ids) => {
                    ExpressionKind::FunctionCall(ident_ref_id, expr_ids)
                }
                AstExpressionKind::While(cond, expr_id) => ExpressionKind::While(cond, expr_id),
                AstExpressionKind::Block(expr_ids) => ExpressionKind::Block(expr_ids),
                AstExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                    ExpressionKind::StructInstantiation(ident_ref_id, fields)
                }
                AstExpressionKind::ArrayInstantiation(values) => {
                    ExpressionKind::ArrayInstantiation(values)
                }
                AstExpressionKind::ArrayAccess(expr, index) => {
                    ExpressionKind::ArrayAccess(expr, index)
                }
                AstExpressionKind::Unary(_, _) => panic!("Cannot convert AstExpressionKind::Unary"),
                AstExpressionKind::Binary(_, _, _) => {
                    panic!("Cannot convert AstExpressionKind::Binary")
                }
                AstExpressionKind::Dummy => panic!("Cannot convert AstExpressionKind::Dummy"),
            },
            typed: Untyped,
        }
    }
}

impl Expression<Typed> {
    pub fn resolved_type_id(&self) -> TypeId {
        self.typed.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    ArrayInstantiation(Vec<ExprId>),
    ArrayAccess(ExprId, ExprId),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}

impl From<AstTarget> for Target {
    fn from(value: AstTarget) -> Self {
        match value {
            AstTarget::Direct(ident_ref_id) => Self::Direct(ident_ref_id),
            AstTarget::Indirect(expr_id) => Self::Indirect(expr_id),
        }
    }
}
