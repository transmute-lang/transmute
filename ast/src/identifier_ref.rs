use crate::identifier::Identifier;
use std::fmt::Debug;
use transmute_core::ids::{IdentId, IdentRefId};
use transmute_core::span::Span;

/// Represents an identifier when used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct IdentifierRef {
    /// ID of this identifier reference
    id: IdentRefId,
    /// The referenced symbol identifier
    ident: Identifier,
}

impl IdentifierRef {
    pub fn new(id: IdentRefId, ident: Identifier) -> Self {
        Self { id, ident }
    }

    pub fn id(&self) -> IdentRefId {
        self.id
    }

    pub fn ident(&self) -> &Identifier {
        &self.ident
    }

    pub fn take_ident(self) -> Identifier {
        self.ident
    }

    pub fn ident_id(&self) -> IdentId {
        self.ident.id()
    }

    pub fn span(&self) -> &Span {
        self.ident.span()
    }
}
