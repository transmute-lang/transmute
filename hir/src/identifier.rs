use crate::bound::{Bound, BoundState, Unbound};
use transmute_ast::identifier::Identifier as AstIdentifier;
use transmute_core::ids::{IdentId, SymbolId};
use transmute_core::span::Span;

/// Represents an identifier when not used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct Identifier<B>
where
    B: BoundState,
{
    id: IdentId,
    span: Span,
    bound: B,
}

impl Identifier<Unbound> {
    pub fn bind(self, symbol_id: SymbolId) -> Identifier<Bound> {
        Identifier::<Bound> {
            id: self.id,
            span: self.span,
            bound: Bound(symbol_id),
        }
    }
}

impl From<AstIdentifier> for Identifier<Unbound> {
    fn from(value: AstIdentifier) -> Self {
        Identifier {
            id: value.id(),
            span: value.take_span(),
            bound: Unbound,
        }
    }
}

impl Identifier<Bound> {
    pub fn symbol_id(&self) -> SymbolId {
        self.bound.0
    }
}

impl<B> Identifier<B>
where
    B: BoundState,
{
    pub fn id(&self) -> IdentId {
        self.id
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}
