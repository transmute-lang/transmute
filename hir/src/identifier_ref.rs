use crate::bound::{Bound, BoundState, Unbound};
use crate::identifier::Identifier;
use std::fmt::Debug;
use transmute_ast::identifier_ref::IdentifierRef as AstIdentifierRef;
use transmute_core::ids::{IdentId, IdentRefId, SymbolId};
use transmute_core::span::Span;

/// Represents an identifier when used as a reference
#[derive(Debug, Clone, PartialEq)]
pub struct IdentifierRef<B>
where
    B: BoundState,
{
    /// ID of this identifier reference
    id: IdentRefId,
    /// The referenced symbol identifier
    ident: Identifier<B>,
}

impl<B> IdentifierRef<B>
where
    B: BoundState,
{
    pub fn id(&self) -> IdentRefId {
        self.id
    }

    pub fn ident(&self) -> &Identifier<B> {
        &self.ident
    }

    pub fn ident_id(&self) -> IdentId {
        self.ident.id()
    }

    pub fn span(&self) -> &Span {
        self.ident.span()
    }
}

impl IdentifierRef<Unbound> {
    pub fn resolved(self, symbol_id: SymbolId) -> IdentifierRef<Bound> {
        IdentifierRef {
            id: self.id,
            ident: self.ident.bind(symbol_id),
        }
    }
}

impl From<AstIdentifierRef> for IdentifierRef<Unbound> {
    fn from(value: AstIdentifierRef) -> Self {
        Self {
            id: value.id(),
            ident: Identifier::from(value.take_ident()),
        }
    }
}

impl IdentifierRef<Bound> {
    pub fn symbol_id(&self) -> SymbolId {
        self.ident.symbol_id()
    }
}
