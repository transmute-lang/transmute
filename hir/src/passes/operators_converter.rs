use bimap::BiMap;
use transmute_ast::expression::{Expression, ExpressionKind};
use transmute_ast::identifier::Identifier;
use transmute_ast::identifier_ref::IdentifierRef;
use transmute_core::ids::{ExprId, IdentId, IdentRefId};
use transmute_core::vec_map::VecMap;

pub struct OperatorsConverter {
    identifiers: BiMap<IdentId, String>,
    identifier_refs: VecMap<IdentRefId, IdentifierRef>,
}

pub struct Output {
    pub identifiers: VecMap<IdentId, String>,
    pub identifier_refs: VecMap<IdentRefId, IdentifierRef>,
    pub expressions: VecMap<ExprId, Expression>,
}

impl OperatorsConverter {
    pub fn new(
        identifiers: VecMap<IdentId, String>,
        identifier_refs: VecMap<IdentRefId, IdentifierRef>,
    ) -> Self {
        Self {
            identifiers: identifiers.into_iter().collect::<BiMap<IdentId, String>>(),
            identifier_refs,
        }
    }

    pub fn convert(mut self, expressions: VecMap<ExprId, Expression>) -> Output {
        let expressions = expressions
            .into_iter()
            .map(|(expr_ir, expr)| (expr_ir, self.convert_expression(expr)))
            .collect::<VecMap<ExprId, Expression>>();

        let identifiers = self
            .identifiers
            .into_iter()
            .collect::<VecMap<IdentId, String>>();

        Output {
            identifiers,
            identifier_refs: self.identifier_refs,
            expressions,
        }
    }

    fn convert_expression(&mut self, expression: Expression) -> Expression {
        match expression.kind() {
            ExpressionKind::Binary(lhs_id, op, rhs_id) => {
                let ident_id = self.ident_id(op.kind().function_name());
                let identifier = Identifier::new(ident_id, op.span().clone());
                let ident_ref_id = IdentRefId::from(self.identifier_refs.len());
                self.identifier_refs
                    .push(IdentifierRef::new(ident_ref_id, identifier));

                Expression::new(
                    expression.id(),
                    ExpressionKind::FunctionCall(ident_ref_id, vec![*lhs_id, *rhs_id]),
                    expression.span().clone(),
                )
            }
            ExpressionKind::Unary(op, expr_id) => {
                let ident_id = self.ident_id(op.kind().function_name());
                let identifier = Identifier::new(ident_id, op.span().clone());
                let ident_ref_id = IdentRefId::from(self.identifier_refs.len());
                self.identifier_refs
                    .push(IdentifierRef::new(ident_ref_id, identifier));

                Expression::new(
                    expression.id(),
                    ExpressionKind::FunctionCall(ident_ref_id, vec![*expr_id]),
                    expression.span().clone(),
                )
            }
            _ => expression,
        }
    }

    fn ident_id(&mut self, function_name: &'static str) -> IdentId {
        if !self.identifiers.contains_right(function_name) {
            let ident_id = IdentId::from(self.identifiers.len());
            self.identifiers.insert(ident_id, function_name.to_string());
        }

        *self
            .identifiers
            .get_by_right(function_name)
            .expect("it is there")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use transmute_ast::lexer::Lexer;
    use transmute_ast::parser::Parser;
    use transmute_ast::pretty_print::Options;
    use transmute_ast::Ast;

    macro_rules! op {
        ($name: ident, $src:expr) => {
            #[test]
            fn $name() {
                let ast = Parser::new(Lexer::new($src)).parse().unwrap();
                let output = OperatorsConverter::new(ast.identifiers, ast.identifier_refs)
                    .convert(ast.expressions);

                let ast = Ast {
                    identifiers: output.identifiers,
                    identifier_refs: output.identifier_refs,
                    expressions: output.expressions,
                    statements: ast.statements,
                    roots: ast.roots,
                };

                let mut w = String::new();
                ast.pretty_print(&Options::default(), &mut w).unwrap();

                assert_snapshot!(w);
            }
        };
    }

    op!(add, "lhs + rhs;");
    op!(sub, "lhs - rhs;");
    op!(mul, "lhs * rhs;");
    op!(div, "lhs / rhs;");
    op!(eq, "lhs == rhs;");
    op!(neq, "lhs != rhs;");
    op!(gt, "lhs > rhs;");
    op!(ge, "lhs >= rhs;");
    op!(lt, "lhs < rhs;");
    op!(le, "lhs <= rhs;");

    op!(minus, "- op;");
}
