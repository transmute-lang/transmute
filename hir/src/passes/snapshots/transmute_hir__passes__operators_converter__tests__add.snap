---
source: hir/src/passes/operators_converter.rs
expression: w
---
add(lhs, rhs);

